lib/posix/pthread_key.c

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <kernel.h>
+#include <misc/dlist.h>
+#include <semaphore.h>
+#include <errno.h>
+#include <posix/pthread.h>
+
+static sem_t sem;
+
+/**
+ * @brief Create a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_create(pthread_key_t *key,
+		void (*destructor)(void *))
+{
+	pthread_key_obj *newKey;
+
+	*key = NULL;
+
+	sem_init(&sem, 0, 1);
+
+	newKey = (pthread_key_obj *)k_malloc(sizeof(pthread_key_obj));
+
+	if (newKey == NULL) {
+		errno = ENOMEM;
+		return -1;
+	}
+
+	sys_dlist_init(&(newKey->KeyDataL));
+
+	newKey->destructor = destructor;
+	*key = (void *)newKey;
+
+	return 0;
+}
+
+/**
+ * @brief Delete a key for thread-specific data
+ *
+ * See IEEE 1003.1
+ */
+int pthread_key_delete(pthread_key_t key)
+{
+	pthread_key_obj *keyObj = (pthread_key_obj *)key;
+	KeyData *keyData;
+
+	sem_wait(&sem);
```

Comment:
For the POSIX implementation, it might be a good idea to use Zephyr objects directly, to reduce the overhead.

Reply:
Okay, I have changed this to use Zephyr `k_sem` objects directly